text/0000-dll-kind.md

```diff
@@ -0,0 +1,103 @@
+
+* If `#[link_ordinal]` is specified the idata section will map from the mangled symbol to the ordinal specified in the dll.
+* If `#[link_name]` is specified the idata section will map from the mangled symbol to the name specified in the dll, without any calling convention decorations added. If calling convention decorations are desired they must be specified explicitly in the value of the `#[link_name]` attribute.
+* If both `#[link_ordinal]` and `#[link_name]` are specified, an error will be emitted.
+* If neither `#[link_ordinal]` nor `#[link_name]` are specified, the idata section will map from the mangled symbol to its unmangled equivalent in the dll. The unmangled symbol will *not* have calling convention decorations.
+
+The idata section that is produced is equivalent to the idata sections found in import libraries, and should result in identical code generation by the linker.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Additional complexity in the language through a new `kind` and a new attribute for specifying ordinals.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The RFC as proposed would allow for full control over linking to symbols from dlls with syntax as close as possible to existing extern blocks.
+
+No alternatives are currently known other than the status quo.
+
+# Prior art
+[prior-art]: #prior-art
+
+Many non-native languages have the ability to import symbols from dlls, but this uses runtime loading by the language runtime and is not the same as what is being proposed here.
+
+Delphi is a native language that has the ability to import symbols from dlls without import libraries.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Bikeshedding on attribute names.
+* Should this feature be extended to other platforms?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+With the features described in this RFC, we would be one step closer towards a fully standalone pure Rust target for Windows that does not rely on any external libraries (aside from the obvious and unavoidable runtime dependence on system libraries), allowing for easy installation and incredibly easy cross compilation.
+
+If that were to happen, we'd no longer need to pretend the pc-windows-gnu toolchain is standalone, and we'd be able to stop bundling MinGW bits entirely in favor of the user's own MinGW installation, thereby resolving a bunch of issues such as [rust-lang/rust#53454](https://github.com/rust-lang/rust/issues/53454).
+
+A future extension of this feature would be the ability to optionally lazily load such external functions, since Rust would naturally have all the information required to do so.
```

Comment:
I think that's a reference to my delayed=true suggestion, e.g.

`#[link(name = "kernel32.dll", kind = "dll", delayed = true)]
`

Reply:
I'd personally prefer the option where the caller of the function chooses whether to do a lazy loaded call of it, and not having to choose at declaration time whether it is lazy loaded. I don't know what the syntax would look like for that.

Reply:
The compiler could still ensure there is only one `LoadLibrary` for a given dll and one `GetProcAddress` for each symbol. Allowing the caller to choose whether they want to lazy load it does not prevent the compiler from ensuring `GetProcAddress` is only called once. If no `dylib` crates are involved, the generation of `GetProcAddress` thunks can be lazily deferred all the way until binary creation time. If `dylib` crates are involved, then either the `dylib` that contains the crate with the declarations would have to generate *all* the thunks, or there could potentially be some duplication across `dylib`s. Since nobody uses `dylib` except `rustc` and people who accidentally used it when they meant to use `cdylib`, it doesn't seem like too big of an issue.

Reply:
As for *why*, it depends on whether the caller is able to deal with the function not existing. If the caller has a fallback, then it can use the lazy loaded versions and fallback if it fails to load. If the caller doesn't have any fallback, then it can use the more efficient static version.

Reply:
@zunzster The user can also write such thunks themselves. It would be preferable for the user to write them since they can choose how to handle the LL/GPA failure if the DLL / function doesn't exist (crash / return a custom `Err()`) rather than have the compiler enforce a specific choice.